examples/permissioned-token/assets/permissions.py

```diff
@@ -0,0 +1,129 @@
+		Txn.rekey_to() == Global.zero_address(),
+		#Txn.applications.length() ==  Int(1), [TEALv3]
+
+		# verify from global state whether controller application passed is the valid one
+		# note: with tealv3 we can just use txn.applications[0] instead of passing an an appArg
+		Btoi(Txn.application_args[1]) == App.globalGet(controller_app_id),
+
+		# first verify correct token value is passed
+		# token_id.hasValue(),
+			# Txn.assets[0] == token_id.value(), [TEALv3]
+
+		# then verify txn sender is the token manager
+		permission_manager.hasValue(),
+		Txn.sender() == permission_manager.value(),
+        )),
+
+        If(
+            # If account is not already whitelisted, update the counter and continue, return otherwise
+            App.localGet(Int(1), Bytes("whitelisted")) == Int(0),
+            App.globalPut(whitelist_count, App.globalGet(whitelist_count) + Int(1)), # only evaluated if above is true
+            Return(Int(1)) # else return
+        ),
+
+        # finally update txn.accounts[1] local state to set whitelisted status as true(1)
+        App.localPut(Int(1), Bytes("whitelisted"), Int(1)),
+        Return(Int(1))
+    ])
+
+    asset_balance = AssetHolding.balance(Int(1), Gtxn[1].xfer_asset())
+    transfer_token = Seq([
+        asset_balance, # load asset_balance of asset_receiver from store
+        Assert(And(
+		Txn.application_args.length() == Int(1),
+		Global.group_size() >= Int(3),
+		Txn.sender() == Gtxn[1].asset_sender(),
+
+		# verify tx.accounts[1] of current_tx should be same as asset receiver
+		Txn.accounts[1] == Gtxn[1].asset_receiver(),
+
+		# rule 1 - check balance of receiver after receiving token <= 100(max_tokens)
+		asset_balance.hasValue(),
+		asset_balance.value() <= App.globalGet(max_tokens),
+
+		# rule 2 - [from, to] accounts must be whitelisted
+		# NOTE: Int(0) == Txn.Sender(), Int(1) == Txn.accounts[1]
+		App.localGet(Int(0), Bytes("whitelisted")) == true, # from account must be whitelisted
+		App.localGet(Int(1), Bytes("whitelisted")) == true  # to account must be whitelisted
+        )),
+        Return(Int(1))
+    ])
+
+    handle_optin = Seq([
+        Assert(And(
+            Txn.application_args.length() == Int(0),
+            Txn.group_index() == Int(0)
+        )),
+        Return(Int(1))
+    ])
+
+    # Verfies that the application_id is 0, jumps to on_deployment.
+    # Verifies that DeleteApplication is used and blocks that call
+    # Verifies that UpdateApplication is used and blocks that call (unsafe for production use).
+    # Verifies that closeOut is used and approves the tx.
+    # Verifies that OptInApplication is used and jumps to handle_optin
+    # Verifies that first argument is "add_whitelist" and jumps to add_whitelist.
+    # Verifies that first argument is "transfer" and jumps to transfer_token.
+    program = Cond(
+        [Txn.application_id() == Int(0), on_deployment],
+        [Txn.on_completion() == OnComplete.UpdateApplication, Return(Int(0))], # block update
```

Comment:
Updates are OK, we want to enable upadates to add more rules if needed in the future.
Let's make it clear here (using a comment) and in the README.

Reply:
Just to confirm: 
+ updating permissions smart contract means updating an existing set of rules/add new rule in current asc
+ adding new rule means add a new rules contract (which we don't support yet, having multiple rule(s) contract(s))

Reply:
yes, but we should be able to deploy a new version of the contract with updated rules.

Reply:
ok, thanks for confirming